streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -0,0 +1,178 @@
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.streams.errors.StreamsException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class StoreChangelogReader implements ChangelogReader {
+    private static final Logger log = LoggerFactory.getLogger(StoreChangelogReader.class);
+
+    private final Consumer<byte[], byte[]> consumer;
+    private final Time time;
+    private final long partitionValidationTimeoutMs;
+    private final Map<String, List<PartitionInfo>> partitionInfo = new HashMap<>();
+    private final Map<TopicPartition, StateRestorer> stateRestorers = new HashMap<>();
+
+
+    public StoreChangelogReader(final Consumer<byte[], byte[]> consumer, final Time time, final long partitionValidationTimeoutMs) {
+        this.consumer = consumer;
+        this.time = time;
+        this.partitionValidationTimeoutMs = partitionValidationTimeoutMs;
+    }
+
+
+    @Override
+    public void validatePartitionExists(final TopicPartition topicPartition, final String storeName) {
+        final long start = time.milliseconds();
+        if (partitionInfo.isEmpty()) {
+            try {
+                partitionInfo.putAll(consumer.listTopics());
+            } catch (final TimeoutException e) {
+                throw new StreamsException(String.format("Could not fetch partition info for topic: %s before expiration of the configured request timeout",
+                                                         topicPartition.topic()));
+            }
+        }
+
+        final long endTime = time.milliseconds() + partitionValidationTimeoutMs;
+        while (!hasPartition(topicPartition) && time.milliseconds() < endTime) {
+            try {
+                final List<PartitionInfo> partitions = consumer.partitionsFor(topicPartition.topic());
+                if (partitions != null) {
+                    partitionInfo.put(topicPartition.topic(), partitions);
+                }
+            } catch (final TimeoutException e) {
+                throw new StreamsException(String.format("Could not fetch partition info for topic: %s before expiration of the configured request timeout",
+                                                         topicPartition.topic()));
+            }
+        }
+
+        if (!hasPartition(topicPartition)) {
+            throw new StreamsException(String.format("Store %s's change log (%s) does not contain partition %s",
+                                                     storeName, topicPartition.topic(), topicPartition.partition()));
+        }
+
+        log.debug("Took {} ms to validate that partition {} exists", time.milliseconds() - start, topicPartition);
+
+    }
+
+    @Override
+    public void register(final StateRestorer restorer) {
```

Comment:
`register` and `validate` functions are quite independent and it seems the only reason we wrap them under the same class is to reuse the `consumer`. How about refactor it a bit as the following:

1. merge the logic of `validate` into `register`, i.e. doing validation while registering.
2. in `StateManager.register`, get rid of the `restoreCallbacks` which is only for the standbyTasks, but just call `logReader.register` for both standby and active tasks, and add a parameter indicating if it is for standby restoration or not.
3. in `StateManager.updateStandbyStates`, access the `logReader` with the given topic-partition, check it is indeed a standby task, and then apply its corresponding `Restorer`'s callback.